clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -158,6 +161,32 @@ public ConsumerCoordinator(GroupRebalanceConfig rebalanceConfig,
         if (autoCommitEnabled)
             this.nextAutoCommitTimer = time.timer(autoCommitIntervalMs);
 
+        // select the rebalance protocol such that:
+        //   1. only consider protocols that are supported by all the assignors. If there is no common protocols supported
+        //      across all the assignors, throw an exception.
+        //   2. if there are multiple protocols that are commonly supported, select the one with the highest id (i.e. the
+        //      id number indicates how advanced the protocol is).
+        // we know there are at least one assignor in the list, no need to double check for NPE
+        if (!assignors.isEmpty()) {
+            List<RebalanceProtocol> supportedProtocols = new ArrayList<>(assignors.get(0).supportedProtocols());
+
+            for (PartitionAssignor assignor : assignors) {
+                supportedProtocols.retainAll(assignor.supportedProtocols());
+            }
+
+            if (supportedProtocols.isEmpty()) {
+                throw new IllegalArgumentException("Specified assignors " +
+                    assignors.stream().map(PartitionAssignor::name).collect(Collectors.toSet()) +
+                    " do not have commonly supported rebalance protocol");
+            }
+
+            Collections.sort(supportedProtocols);
+
+            protocol = supportedProtocols.get(supportedProtocols.size() - 1);
+        } else {
+            protocol = null;
```

Comment:
Won't this break in the `switch` below? Am I missing something?

Reply:
We have a check at the beginning of `ConsumerCoordinator.poll`:

```
            if (protocol == null) {
                throw new IllegalStateException("User confingure ConsumerConfig#PARTITION_ASSIGNMENT_STRATEGY_CONFIG to empty " +
                    "while trying to subscribe for group protocol to auto assign partitions");
            }
```

Basically if user does not call `subscribe()` it is okay to not set the protocol at all; otherwise we will throw error.

Reply:
Ok. We also have a check in `subscribe` to ensure that the set of assignors is not empty. There might be a way to remove the redundant checking.

By the way, there's a typo above: `confingure`.

Reply:
Ack.

Reply:
Yeah I think we can indeed remove it:

1. if no assignors specified, then `subscribe` will throw.
2. if assignors specified but no common supported protocol, then constructor will throw.